app/assets/javascripts/govuk_publishing_components/components/modal-dialogue.js

```diff
@@ -0,0 +1,92 @@
+window.GOVUK = window.GOVUK || {}
+window.GOVUK.Modules = window.GOVUK.Modules || {};
+
+(function (Modules) {
+  function ModalDialogue () { }
+
+  ModalDialogue.prototype.start = function ($module) {
+    this.$module = $module[0]
+    this.$dialogBox = this.$module.querySelector('.gem-c-modal-dialogue__box')
+    this.$closeButton = this.$module.querySelector('.gem-c-modal-dialogue__close-button')
+    this.$body = document.querySelector('body')
+
+    this.$module.open = this.handleOpen.bind(this)
+    this.$module.close = this.handleClose.bind(this)
+    this.$module.focusDialog = this.handleFocusDialog.bind(this)
+    this.$module.boundKeyDown = this.handleKeyDown.bind(this)
+
+    var $triggerElement = document.querySelector(
+      '[data-toggle="modal"][data-target="' + this.$module.id + '"]'
+    )
+
+    if ($triggerElement) {
+      $triggerElement.addEventListener('click', this.$module.open)
+    }
+
+    if (this.$closeButton) {
+      this.$closeButton.addEventListener('click', this.$module.close)
+    }
+  }
+
+  ModalDialogue.prototype.handleOpen = function (event) {
+    if (event) {
+      event.preventDefault()
+    }
+
+    this.$body.classList.add('app-o-template__body--modal')
+    this.$body.classList.add('app-o-template__body--blur')
+    this.$focusedElementBeforeOpen = document.activeElement
+    this.$module.style.display = 'block'
+    this.$dialogBox.focus()
+
+    document.addEventListener('keydown', this.$module.boundKeyDown, true)
+  }
+
+  ModalDialogue.prototype.handleClose = function (event) {
+    if (event) {
+      event.preventDefault()
+    }
+
+    this.$body.classList.remove('app-o-template__body--modal')
+    this.$body.classList.remove('app-o-template__body--blur')
+    this.$module.style.display = 'none'
+    this.$focusedElementBeforeOpen.focus()
+
+    document.removeEventListener('keydown', this.$module.boundKeyDown, true)
+  }
+
+  ModalDialogue.prototype.handleFocusDialog = function () {
+    this.$dialogBox.focus()
+  }
+
+  // while open, prevent tabbing to outside the dialogue
+  // and listen for ESC key to close the dialogue
+  ModalDialogue.prototype.handleKeyDown = function (event) {
+    var KEY_TAB = 9
+    var KEY_ESC = 27
+
+    switch (event.keyCode) {
+      case KEY_TAB:
+        if (event.shiftKey) {
+          if (document.activeElement === this.$dialogBox) {
+            event.preventDefault()
+            this.$closeButton.focus()
+          }
+        } else {
+          if (document.activeElement === this.$closeButton) {
+            event.preventDefault()
+            this.$dialogBox.focus()
+          }
+        }
+
+        break
+      case KEY_ESC:
+        this.$module.close()
+        break
+      default:
+        break
+    }
+  }
+
+  Modules.ModalDialogue = ModalDialogue
```

Comment:
It looks like this code has been modified slightly to fit with the GOVUK.Modules structure, but it's still internally using the previous conventions (prototype, for example). What's the benefit of this?

My concern is that this component's JS is structured slightly differently from all the others unnecessarily (I like consistency). Obviously rewriting it at this point is a bit of a pain, but we should have a conversation about the best way of writing these modules at some point.

Reply:
The main difference between using prototype vs defining methods straight in the constructor is in performance. Assigning methods within the constructor creates a new function object for every class object, using more memory and taking more processing time when calling the class constructor.

[This repository's conventions](https://github.com/alphagov/govuk_publishing_components/blob/master/docs/component_conventions.md#javascript) regarding JavaScript aim to use the GOV.UK Frontend conventions (I like consistency too), which means that the legacy JavaScript code should be updated. Meanwhile the JavaScript code we introduce should stick to the conventions we aim for.